proto/tap.proto

```diff
@@ -21,6 +21,9 @@ message ObserveRequest {
   // Encodes request-matching logic.
   Match match = 2;
 
+  // Conditionally adds headers and trailers to requests and responses in tap events.
+  bool includeMetadata = 3;
```

Comment:
nit: proto field names should be `lower_snake_case` (as elsewhere).

I think _metadata_ is a confusing term here: _all_ of the stuff emitted in tap events is metadata (and, the type we return is called `Headers`).

Is it appropriate for this field to be on the top-level tap request message? Consider what happens as the proxy grows support for tapping additional protocols. Would this flag be HTTP-specific? or should it apply to others?  Might we want to express additional configuration about how headers are returned from the proxy in the future?  I think we need to carve out a bit more API space for ourselves to grow here (for instance, as tap bodies, etc, are added).

Note how we express matches in a "protomorphic" fashion (get it? protobuf + polymorphism. i just made that up):
```protobuf

  message Match {
    message Seq {
      repeated Match matches = 1;
    }

    oneof match {
      Seq all = 1;
      Seq any = 2;
      Match not = 3;

      Tcp source  = 4;
      Tcp destination = 5;
      Http http = 6;

      Label destination_label = 7;
      Label route_label = 8;
    }
   ...

    message Http {
      oneof match {
        http_types.Scheme     scheme    = 1;
        http_types.HttpMethod method    = 3;
        StringMatch           authority = 2;
        StringMatch           path      = 4;
        // TODO Header        header    = 4;
      }
    ...
```

We should probably do a similar (though much simpler version of this to give ourselves clean API space moving forward. Perhaps something like the following:

```protobuf
    repeated Extract extracts = 3;
    ...
    message Extract {
      oneof extract {
        Http http = 1;
      }

       message Http {
         oneof extract {
           Headers headers = 1;
         }

         message Headers {}
       }
    }
```

In this case, the proxy would only emit headers if there is an Extract message containing an Http message containing an (empty) Headers message. But, while this is definitely more verbose, it has a few nice properties:
- HTTP-specific logic isn't leaking out into the core of the tap request -- it's just one type of tap that we'll ultimately be able to execute.
- We have an obvious place to limit which headers are extracted, if we care to do so.
- HTTP payload extraction can follow similarly

I'm not sure if extract is the right word for this, but it feels congruous with _match_.

Reply:
Thanks this is a really good point. It is more verbose, but as this is part of the proxy API I think it's a good thing to use initially.

I think extract and include are both good options. Extract seems a little more succinct in terms of readability on what this protobuf message is doing.